android/app/src/main/java/com/criticalblue/auth/demo/auth/AuthRepo.java

```diff
@@ -247,83 +274,55 @@ private void startUserAuth() {
     }
 
     public void notifyUserAgentResponse(Intent data, int returnCode) {
-        if (returnCode != app.RC_AUTH) {
+        if (returnCode != BooksApp.RC_AUTH) {
             failLogin(new AuthException("User authorization was cancelled"));
             return;
         }
 
         AuthorizationResponse resp = AuthorizationResponse.fromIntent(data);
         AuthorizationException ex = AuthorizationException.fromIntent(data);
+
         if (resp == null) {
             failLogin(new AuthException("User authorization failed"));
-            return;
         } else {
             authState.update(resp, ex);
             finishUserAuth();
         }
     }
 
     private void finishUserAuth() {
-        Log.i(TAG, "Finishing user auth");
+        Log.i(TAG, "---> Finishing user auth");
 
         loginListener.onEvent(AuthRepo.this, AUTH_USER_AUTH_FINISH);
 
-        startApproovCheck();
-    }
-
-    private void startApproovCheck() {
-        Log.i(TAG, "Starting Approov check");
-
-        ApproovAttestation.shared().fetchApproovToken(new ApproovTokenFetcher(),
-                "www.googleapis.com"); // checking google api domain rather than
-                                          // self-signed oauth2 adapter
-    }
-
-    private String approovToken = "BAD";
-
-    public class ApproovTokenFetcher implements TokenInterface {
-
-        ApproovTokenFetcher() { /* EMPTY */ }
-
-        public void approovTokenFetchResult(final ApproovResults approovResults) {
-
-            ApproovAttestation.AttestationResult result = approovResults.getResult();
-            approovToken = approovResults.getToken();
-            Log.i(TAG, "Approov token = \"" + approovToken + "\"");
-
-            if (result == ApproovAttestation.AttestationResult.SUCCESS){
-                finishApproovCheck();
-            } else if (result == ApproovAttestation.AttestationResult.FAILURE){
-                failLogin(new AuthException("Approov checking failed"));
-                return;
-            }
-        }
-    }
-
-    private void finishApproovCheck() {
-        Log.i(TAG, "Finishing Approov check");
-
         startCodeExchange();
     }
 
     private void startCodeExchange() {
-        Log.i(TAG, "Starting code exchange");
+        Log.i(TAG, "---> Starting code exchange");
 
         loginListener.onEvent(AuthRepo.this, AUTH_CODE_EXCHANGE_START);
 
         AuthorizationResponse resp = authState.getLastAuthorizationResponse();
 
-        // use Approov token as secret
+        String clientSecret = fetchJustInTimeClientSecret();
 
-        ClientAuthentication clientAuth = new ClientSecretPost(approovToken);
-        approovToken="BAD";
+        if (clientSecret == null) {
+            failLogin(new AuthException("Failed to fetch the client_secret. Did you registered the APK with Approov?"));
```

Comment:
You will have called failLogin already, It might actually be cleaner to essentially inline the fetchJustInTmeClientSecret here

Reply:
Or get fetchJustInTimeClientSecret to throw an AuthException, which you can catch and pass to failLogin here in one place.

Reply:
I will move the check inside the `fetchJustInTimeClientSecret()` and inline the call.

Reply:
Solved in the last commit.